### PR TITLE
Only build host libraries with lto

### PR DESCRIPTION
**Pull request by gottesmm**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci Please python lint

**Comment by gottesmm** [comment]
@swift-ci Please smoke test

File: CMakeLists.txt [hunk 1]
```diff
@@ -140,8 +140,9 @@ set(SWIFT_NATIVE_CLANG_TOOLS_PATH "" CACHE STRING
 set(SWIFT_NATIVE_SWIFT_TOOLS_PATH "" CACHE STRING
     "Path to the directory that contains Swift tools that are executable on the build machine")
 
-option(SWIFT_TOOLS_ENABLE_LTO
-  "If set to true, build the swift compiler with link time optimization enabled" FALSE)
+set(SWIFT_TOOLS_ENABLE_LTO OFF CACHE STRING "Build Swift tools with LTO. One
```
Comment (gribozavr) [inline CMakeLists.txt]: I'd say something like "This option only affects the tools that run on the host (the compiler), and has no effect on the target libraries (the standard library and the runtime)."

File: cmake/modules/AddSwift.cmake [hunk 2]
```diff
@@ -53,6 +53,17 @@ function(compute_library_subdir result_var_name sdk arch)
   set("${result_var_name}" "${SWIFT_SDK_${sdk}_LIB_SUBDIR}/${arch}" PARENT_SCOPE)
 endfunction()
 
+function(_compute_lto_flag option out_var)
+  string(TOLOWER "${option}" lowercase_option)
+  if (lowercase_option STREQUAL "full")
+    set(${out_var} "-flto=full" PARENT_SCOPE)
+  elseif (lowercase_option STREQUAL "thin")
+    set(${out_var} "-flto=thin" PARENT_SCOPE)
```
Comment (gribozavr) [inline cmake/modules/AddSwift.cmake]: Thank you for supporting `=thin`!

File: cmake/modules/AddSwift.cmake [hunk 3]
```diff
@@ -110,14 +121,15 @@ function(_add_variant_c_compile_link_flags)
   endif()
 
   if(CFLAGS_ENABLE_LTO)
-    list(APPEND result "-flto")
+    _compute_lto_flag(${CFLAGS_ENABLE_LTO} _lto_flag_out)
+    list(APPEND result "${_lto_flag_out}")
   endif()
```
Comment (gribozavr) [inline cmake/modules/AddSwift.cmake]: You could move the if statement into `_compute_lto_flag` (but then make sure to remove the double quotes from the append command).

File: cmake/modules/AddSwift.cmake [hunk 4]
```diff
@@ -797,12 +809,18 @@ function(_add_swift_library_single target name)
     set(enable_assertions "${LLVM_ENABLE_ASSERTIONS}")
     set(analyze_code_coverage "${SWIFT_ANALYZE_CODE_COVERAGE}")
   endif()
+
+  if (SWIFTLIB_SINGLE_HOST_LIBRARY)
```
Comment (gribozavr) [inline cmake/modules/AddSwift.cmake]: The name of the flag is probably misleading.  It means "no matter what the defaults say, also build this library for the host".  So if the defaults say that this library should be built for the host, then this flag won't be set -- which is the case for all compiler libraries.  Maybe this flag should be called something like `FORCE_BUILD_FOR_HOST_SDK`.

The condition you need to check here is `NOT SWIFTLIB_SINGLE_TARGET_LIBRARY `.

File: utils/build-script [hunk 5]
```diff
@@ -1468,10 +1471,10 @@ details of the setups of other systems or automated environments.""")
                 pipes.quote(opt) for opt in args.extra_cmake_options)
         ]
 
-    if args.lto:
+    if args.lto_type is not None:
         build_script_impl_args += [
-            "--llvm-enable-lto",
-            "--swift-tools-enable-lto"
+            "--llvm-enable-lto=%s" % args.lto_type,
+            "--swift-tools-enable-lto=%s" % args.lto_type
```
Comment (gribozavr) [inline utils/build-script]: I like the symmetry here 👍

File: utils/build-script-impl [hunk 6]
```diff
@@ -74,8 +74,8 @@ KNOWN_SETTINGS=(
     swift-build-type            "Debug"          "the CMake build variant for Swift"
     swift-enable-assertions     "1"              "enable assertions in Swift"
     swift-analyze-code-coverage "not-merged"     "Code coverage analysis mode for Swift (false, not-merged, merged). Defaults to false if the argument is not present, and not-merged if the argument is present without a modifier."
-    swift-tools-enable-lto      "0"              "enable LTO compilation of Swift tools. *NOTE* This does not include the swift standard library and runtime."
-    llvm-enable-lto             "0"              "enable LTO compilation of LLVM/Clang."
+    swift-tools-enable-lto      ""               "enable LTO compilation of Swift tools. *NOTE* This does not include the swift standard library and runtime. Must be set to one of thin or full"
+    llvm-enable-lto             ""               "Must be set to one of thin or full"
```
Comment (gribozavr) [inline utils/build-script-impl]: Could you put "thin" and "full" in quotes to make it easier to read?  (here and in other help messages)

**Comment by gribozavr** [comment]
LGTM modulo comments!

File: cmake/modules/AddSwift.cmake (hunk 4 above)
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: Yes. I will rename the flag as apart of this stream of commits.

**Comment by gottesmm** [comment]
@swift-ci please smoke test

**Comment by gottesmm** [comment]
I had to make some more changes and I need to chop c8a1a85 in two. But for a second I am using the PR to do some quick smoke testing.

**Comment by gottesmm** [comment]
@swift-ci Please smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci Please python lint

**Pull request merged**